File: ocaml/01-10/04.ml
(* Find the number of elements of a list. (easy) *)

(* len : int -> 'a list -> int *)
let rec len acc = function
  | [] -> acc
  | x::xs -> len (acc + 1) xs

(* length : 'a list -> int *)
let length ls = len 0 ls
 
let _ = assert ((length [ "a" ; "b" ; "c"]) = 3)
let _ = assert ((length []) = 0)

File: prolog/01-10/04.pl
/*
P04 (*) Find the number of elements of a list.
*/

cacc([], Acc, Z) :- Z = Acc.
cacc([_|T], Acc, Z) :- Acc1 is Acc + 1, cacc(T, Acc1, Z).

count([], 0).
count(Xs, N) :- cacc(Xs, 0, N).
